vaadin-combo-box-flow-parent/vaadin-combo-box-flow/src/main/resources/META-INF/resources/frontend/comboBoxConnector.js

```diff
@@ -2,285 +2,276 @@ import { Debouncer } from '@polymer/polymer/lib/utils/debounce.js';
-      // and submitting the filtered items to specified callback.
-      // The filter used is the one from combobox, not the lastFilter stored since
-      // that may not reflect user's input.
-      const performClientSideFilter = tryCatchWrapper(function (page, filter, callback) {
-        let filteredItems = page;
-
-        if (filter) {
-          filteredItems = page.filter((item) => comboBox.$connector.filter(item, filter));
-        }
-
-        callback(filteredItems, filteredItems.length);
-      });
-
-      // Prevent setting the custom value as the 'value'-prop automatically
-      comboBox.addEventListener(
-        'custom-value-set',
-        tryCatchWrapper((e) => e.preventDefault())
-      );
+        comboBox.dataProvider(params, callback);
+      } else if (rangeMax - rangeMin + 1 !== activePages.length) {
+        // Wasn't a sequential page index, clear the cache so combo-box will request for new pages
+        clearPageCallbacks();
+      } else {
+        // The requested page was sequential, extend the requested range
+        const startIndex = params.pageSize * rangeMin;
+        const endIndex = params.pageSize * (rangeMax + 1);
 
-      comboBox.itemClassNameGenerator = tryCatchWrapper(function (item) {
-        return item.className || '';
-      });
-    })(comboBox)
-};
+        serverFacade.requestData(startIndex, endIndex, params);
+      }
+    }
+  };
+
+  comboBox.$connector.clear = (start, length) => {
+    const firstPageToClear = Math.floor(start / comboBox.pageSize);
+    const numberOfPagesToClear = Math.ceil(length / comboBox.pageSize);
+
+    for (let i = firstPageToClear; i < firstPageToClear + numberOfPagesToClear; i++) {
+      delete cache[i];
+    }
+  };
+
+  comboBox.$connector.filter = (item, filter) => {
+    filter = filter ? filter.toString().toLowerCase() : '';
+    return comboBox._getItemLabel(item, comboBox.itemLabelPath).toString().toLowerCase().indexOf(filter) > -1;
+  };
+
+  comboBox.$connector.set = (index, items, filter) => {
+    if (filter != serverFacade.getLastFilterSentToServer()) {
+      return;
+    }
+
+    if (index % comboBox.pageSize != 0) {
+      throw 'Got new data to index ' + index + ' which is not aligned with the page size of ' + comboBox.pageSize;
+    }
+
+    if (index === 0 && items.length === 0 && pageCallbacks[0]) {
+      // Makes sure that the dataProvider callback is called even when server
+      // returns empty data set (no items match the filter).
+      cache[0] = [];
+      return;
+    }
+
+    const firstPageToSet = index / comboBox.pageSize;
+    const updatedPageCount = Math.ceil(items.length / comboBox.pageSize);
+
+    for (let i = 0; i < updatedPageCount; i++) {
+      let page = firstPageToSet + i;
+      let slice = items.slice(i * comboBox.pageSize, (i + 1) * comboBox.pageSize);
+
+      cache[page] = slice;
+    }
+  };
+
+  comboBox.$connector.updateData = (items) => {
```

Comment:
nit: some functions are changed to arrow functions and others are not. 
Maybe we could align these in a separate PR.